### DISCUSSION TITLE
Drop Support for Qt `<5.15.2`

### DISCUSSION THREAD

**Discussion by Nerixyz** [Chatterino/chatterino2#4292] [Ideas] [answered]
# Reasoning

The main reason to keep support for Qt 5.12.12, in my understanding, is/was to have an easy path for users of Ubuntu 20.04 using the `.deb` artifact.

Now, all up-to-date distributions of Linux mentioned in [`BUILDING_ON_LINUX.md`](https://github.com/Chatterino/chatterino2/blob/master/BUILDING_ON_LINUX.md) provide Qt `^5.15.2`. Users of Windows and macOS can update Qt using the official Qt installer (or additionally brew in the case of macOS).

Supporting Qt 5.12.12 is slowing down development while providing little to no value. This mainly shows in the widely `QString`s which got their specialization for `std::hash` in 5.14, thus `QStringHash` has to be included everywhere a std-container like `std::unordered_map` or `std::unordered_set` with a `QString` as a key is used. Furthermore, some important [`QStringView`](https://doc.qt.io/qt-5/qstringview.html) methods were added in 5.15.2, which is needed for Qt6 compatibility, as `QStringRef` was removed.

I think it's worth considering dropping support for these older Qt versions.

# OS Support
## Windows

Qt Installer only shows the latest version (5.15.2).

## Mac

Qt Installer only shows the latest version.

Through brew: [5.15.8](https://formulae.brew.sh/formula/qt@5)

The build instructions should be updated to mention that 5.15.2 is required.

### Ubuntu

| Package | Focal (20.04) <br> EOS Apr-2025 | Jammy (22.04) <br> EOS Apr-2027 |
|---|---|---|
| `qttools-opensource-src` ⇾  `qttools5-dev` | [5.12.8-0ubuntu1](https://packages.ubuntu.com/focal/qttools5-dev) | [5.15.3-1](https://packages.ubuntu.com/jammy/qttools5-dev) |
| `qtmultimedia-opensource-src` ⇾ `qtmultimedia5-dev` | [5.12.8-0ubuntu1](https://packages.ubuntu.com/focal/qtmultimedia5-dev) | [5.15.3-1](https://packages.ubuntu.com/jammy/qtmultimedia5-dev) |
| `qtimageformats-opensource-src` ⇾ `qt5-image-formats-plugins` | [5.12.8-0ubuntu1](https://packages.ubuntu.com/focal/qt5-image-formats-plugins) | [5.15.3-1](https://packages.ubuntu.com/jammy/qt5-image-formats-plugins) |
| `qtsvg-opensource-src` ⇾ `libqt5svg5-dev`  | [5.12.8-0ubuntu1](https://packages.ubuntu.com/focal/libqt5svg5-dev) | [5.15.3-1](https://packages.ubuntu.com/jammy/libqt5svg5-dev) |

However, you can install Qt 5.15.2 on Focal and earlier using either the [Qt online installer](https://www.qt.io/download-qt-installer) or [PPAs](https://launchpad.net/~beineri). It was [mentioned](https://github.com/Chatterino/chatterino2/pull/3573#pullrequestreview-894481145), that the last LTS is supported.

The build instructions should be updated to mention Jammy (22.04) as the minimum (suggested) version.

### Arch Linux

Uses Qt 5.15.2.

### Fedora

Fedora 35 (Nov-21) provides 5.15.2 - 36 (May-22) and 37 (Nov-22) provide 5.15.3 [unsure about older versions].

The build instructions should be updated to mention 35 as the minimum (known) version.

### NixOS

NixOS provides in Qt 5.12.3 in 22.05+ [unsure about older versions].

The build instructions should be updated to mention 22.05 as the minimum (known) version.

**Comment by pajlada** [comment]
Requirements
 - Split up Ubuntu CI to build on 20.04 & 22.04 (and provide 20.04-specific and 22.04-specific deb builds)

**Comment by Nerixyz** [comment]
With Qt 6.5 scheduled to release on the 3rd of April. I think the CI can be updated to add **experimental** builds with Qt 6 (this discussion seems to be a decent place to discuss this).

Since Ubuntu 20.04 requires 5.12, this version can't be dropped there. However, it can be dropped on Windows and macOS. I propose the following builds in CI:

| Qt Version | FreeBSD | macOS | Ubuntu | Windows |
|---|:---:|:---:|:---:|:---:|
| 5.12.12 | ❌  | ❌ | ✅  | ❌ |
| 5.15.2 | ✅ | ✅  | ✅  | ✅ |
| 6.2.4 | ? |  ? | ✅  | ? |
| 6.5.0 | ? | ✅ | ❌ | ✅ |

It would be fine for now to only build on 6.2.4 and ignore the latest version. 6.2.4 is needed because that's what Ubuntu 22.04 uses. You can read what's new in each version [here](https://doc.qt.io/qt-6/whatsnewqt6.html). The CI build artifacts would be marked "experimental".

### Status on known Qt 6 bugs in Chatterino

There are no more major bugs with Qt 6.

**Reply by LosFarmosCTL** [reply]
> Status on known Qt 6 bugs in Chatterino

Adding to this, Qt 6 would enable us to build M1 versions for macOS.

This fixes #4445 and #2300

**Reply by pajlada** [reply]
I don't think M1 versions are that simple, we still require CI to support it & building with Qt 5.15 on M1 already seems to work

**Reply by pajlada** [reply]
To respond to Nerixyz's comment: This version grid looks good to me

**Reply by LosFarmosCTL** [reply]
> I don't think M1 versions are that simple, we still require CI to support it & building with Qt 5.15 on M1 already seems to work

There would have to be changes to the CI yeah, but it doesn't require GitHub to support Apple Silicon or anything.

While 5.15 can compile an Apple Silicon binary, the big advantage of Qt 6 (and commercial 5.15.19) is that it supports universal binaries, meaning there can just be one version thats compiled on Intel and works for both platforms.

**Reply by iProdigy** [reply]
Another reason to work towards Qt 6.5.0: OpenSSL 1.1.1 is [EOL](https://www.openssl.org/blog/blog/2023/03/28/1.1.1-EOL/) on 2023-09-11 (& only 6.5+ are built against OpenSSL 3)

**Comment by Nerixyz** [comment]
As a test, I removed Qt 5.12 support in https://github.com/Nerixyz/chatterino2/commit/0e2fdffe1f5945815da049fd1d2e57c4981e0c09. This removes the `deb` artifact on Ubuntu 20.04. The AppImage as well as the Flatpak however both contain Qt 5.15 (this would [probably] make it possible to move to Qt 6 as well).

Additionally, I updated GCC to version 10 for better C++ 20 compatibility.

---

Since opening this discussion, OBS also announced dropping support for Ubuntu 20.04 (and Qt 5) in https://github.com/obsproject/obs-studio/discussions/9055.

**Comment by Nerixyz** [comment, marked as answer]
🎉 It's done, yay! 

<sub>Sidenote because I can't find a good place: `std::map` and `std::unordered_map` support [heterogeneous lookup](https://www.cppstories.com/2021/heterogeneous-access-cpp20/), maybe we could use that in some cases (e.g. LinkParser could profit).</sub>
